intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java

```diff
@@ -0,0 +1,167 @@
+package saros.intellij.ui.views.buttons;
+
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.ui.InputValidator;
+import com.intellij.openapi.util.TextRange;
+import java.util.function.BiPredicate;
+import org.apache.log4j.Logger;
+import org.jetbrains.annotations.NotNull;
+import saros.SarosPluginContext;
+import saros.communication.connection.ConnectionHandler;
+import saros.communication.connection.IConnectionStateListener;
+import saros.exceptions.IllegalAWTContextException;
+import saros.exceptions.OperationCanceledException;
+import saros.intellij.ui.Messages;
+import saros.intellij.ui.util.IconManager;
+import saros.intellij.ui.util.SafeDialogUtils;
+import saros.net.ConnectionState;
+import saros.net.xmpp.JID;
+import saros.net.xmpp.contact.XMPPContactsService;
+import saros.repackaged.picocontainer.annotations.Inject;
+
+/** Button to add a contact to the currently connected account. */
+public class AddContactButton extends AbstractToolbarButton {
+  private static final Logger log = Logger.getLogger(AddContactButton.class);
+
+  @Inject private XMPPContactsService xmppContactsService;
+  @Inject private ConnectionHandler connectionHandler;
+
+  private final IConnectionStateListener connectionStateListener =
+      (state, error) -> {
+        if (state == ConnectionState.CONNECTED) {
+          setEnabled(true);
+        } else {
+          setEnabled(false);
+        }
+      };
+
+  public AddContactButton(@NotNull Project project) {
+    super(project, null, Messages.AddContactButton_tooltip, IconManager.ADD_CONTACT_ICON);
+
+    SarosPluginContext.initComponent(this);
+
+    connectionHandler.addConnectionStateListener(connectionStateListener);
+
+    addActionListener(actionEvent -> showAddContactDialog());
+
+    if (!connectionHandler.isConnected()) {
+      setEnabled(false);
+    }
+  }
+
+  @Override
+  public void dispose() {
+    connectionHandler.removeConnectionStateListener(connectionStateListener);
+  }
+
+  /**
+   * Shows the dialogs requesting the JID and (optional) nickname for the account to add as a
+   * contact.
+   *
+   * <p>Calls {@link XMPPContactsService#addContact(JID, String, BiPredicate)} if the dialogs were
+   * not canceled.
+   */
+  private void showAddContactDialog() {
+
+    // TODO better communicate issue to user
+    // TODO use checkInput as it disables the OK button instead of just doing nothing when OK is
+    //  clicked on illegal input.
+    InputValidator inputValidator =
+        new InputValidator() {
+          @Override
+          public boolean checkInput(String inputString) {
+            return true;
+          }
+
+          @Override
+          public boolean canClose(String inputString) {
+            JID jid = new JID(inputString);
+
+            return jid.isValid() && !jid.getName().isEmpty();
+          }
+        };
+
+    String userID;
+
+    try {
```

Comment:
This is not very user friendly. First you ask for the JID and afterwards another dialog is opened that asks for an optional nickname ? Any reasons why this is different from Eclipse where you can insert the JID and nickname in the same wizard/dialog page ?

Reply:
This is how the current user dialog is mostly handled (account addition works this way as well; the incoming session and project negotiation are the only components that currently use the wizard).

Theoretically, this could be implemented using the Saros/I wizard, but the current implementation (located [here](https://github.com/saros-project/saros/tree/master/intellij/src/saros/intellij/ui/wizards) is not very good from what I can tell. It
- does not use the IntelliJ swing wrappers consistently,
- is somewhat difficult to work with when implementing new pages (as basic elements of the wizard, like control over the buttons, are not accessible to the pages (as far as I know), making it hard to provide a well-rounded user interaction), and
- lacks basic UI features like "automatic window scaling to content size" and "content size scaling to window size" (i.e. it plain looks bad and most of the time it does not have the right size for its displayed content, leading to cut-off text or a *lot* of blank space).

So for the quick and dirty implementation of the add contact dialog, I chose to use the default dialogs.

As I already mentioned multiple times, the UI is still one of the most outdated/broken/lackluster aspects of the Saros/I implementation and would need a lot more attention. But, as it *kind of* works currently, it seems like a lower priority than providing basic Saros features to me (and I am not used to/don't like implementing GUI logic 😉).

And even with the lacking quality of the existing components, when working on the GUI, I would still put implementing new features (like the settings) before improving the existing ones (unless they are related). Either way, I will create issues for the missing GUI features in Saros/I (and will probably also include the stuff that needs improvements). Then we can maybe get some outside collaborators (or students) to work on that as it should mostly be independent from the main Saros logic, meaning not that much base knowledge of Saros is required (so it is easier to get going) and the influence on other components is limited.